server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutorService.java

```diff
@@ -230,6 +375,287 @@ private void abortMergeTask(MergeTask mergeTask) {
+        }
+    }
+
+    static class BudgetTracker<E> {
+        private final ToLongFunction<? super E> budgetFunction;
+        private long availableBudget;
+        private final IdentityHashMap<E, Long> commitedBudgetPerElement;
+        private final List<E> budgetOverflow;
+        private final Consumer<? super E> budgetOverflowConsumer;
+        private final Object mutex = new Object();
+
+        BudgetTracker(ToLongFunction<? super E> budgetFunction, long availableBudget, Consumer<? super E> budgetOverflowConsumer) {
+            this.budgetFunction = budgetFunction;
+            this.availableBudget = availableBudget;
+            this.commitedBudgetPerElement = new IdentityHashMap<>();
+            this.budgetOverflow = new ArrayList<>();
+            this.budgetOverflowConsumer = budgetOverflowConsumer;
+        }
+
+        Releasable checkBudget(E element) throws BudgetOverflowException {
+            long elementBudget = budgetFunction.applyAsLong(element);
+            synchronized (mutex) {
+                if (elementBudget > availableBudget) {
+                    budgetOverflow.add(element);
+                    throw new BudgetOverflowException();
+                } else {
+                    assert commitedBudgetPerElement.containsKey(element) == false;
+                    commitedBudgetPerElement.put(element, elementBudget);
+                    availableBudget -= elementBudget;
+                    assert availableBudget > 0;
+                    return () -> {
+                        synchronized (mutex) {
+                            assert commitedBudgetPerElement.containsKey(element);
+                            availableBudget += commitedBudgetPerElement.remove(element);
+                            dropBudgetOverflow();
+                        }
+                    };
+                }
+            }
+        }
+
+        void updateBudget(long availableBudget) {
+            assert availableBudget > 0;
+            synchronized (mutex) {
+                long previouslyAvailableBudget = this.availableBudget;
+                this.availableBudget = availableBudget;
+                // update the per-element budget
+                commitedBudgetPerElement.replaceAll((e, v) -> budgetFunction.applyAsLong(e));
+                // update available budget given the per-element budget
+                this.availableBudget -= commitedBudgetPerElement.values().stream().reduce(0L, Long::sum);
+                if (this.availableBudget > previouslyAvailableBudget) {
+                    dropBudgetOverflow();
+                }
+            }
+        }
+
+        private void dropBudgetOverflow() {
+            synchronized (mutex) {
+                for (E element : budgetOverflow) {
+                    budgetOverflowConsumer.accept(element);
+                }
+                budgetOverflow.clear();
+            }
+        }
```

Comment:
As soon as more budget becomes available, re-enqueue all the merge tasks that were over-budget.
It's complicated to re-enqueue only "within" available budget.